### PR TITLE
Add API Protect

### PR DESCRIPTION
**Pull request by sargun**
API protect, when optionally enabled, is a feature that allows
for protection of automatically generated AWS credentials
by bounding them to a given host (VPC, NAT gateways).

### CODE REVIEW

File: cmd/titus-metadata-service/main.go [hunk 1]
```diff
@@ -146,6 +162,21 @@ func main() {
 			EnvVar:      types.TitusMetatronVariableName,
 			Destination: &metatronEnabled,
 		},
+		cli.StringFlag{
+			Name:        "vpc-id",
+			EnvVar:      "EC2_VPC_ID",
+			Destination: &vpcID,
+		},
+		cli.StringFlag{
```
Comment (sargun) [inline cmd/titus-metadata-service/main.go]: TODO: Remove.

File: cmd/titus-metadata-service/main.go [hunk 2]
```diff
@@ -158,23 +189,43 @@ func main() {
...
+
 		if metatronEnabled {
 			log.Info("Metatron enabled!")
-			if signer, err = identity.NewDefaultSigner(); err != nil {
-				log.Fatal(err)
+			if signer, err := identity.NewDefaultSigner(); err != nil {
+				log.WithError(err).Fatal("Cannot instantiate new default signer")
+				mdscfg.Signer = signer
+			}
+			if container, err := readTaskConfigFile(titusTaskInstanceID); err != nil {
```
Comment (sargun) [inline cmd/titus-metadata-service/main.go]: @rgulewich I hope it was okay to hoist this in here.

File: cmd/titus-metadata-service/main.go [hunk 3]
```diff
@@ -158,23 +189,43 @@ func main() {
...
+				log.WithError(err).Fatal("Cannot read container config file")
+				mdscfg.Container = container
 			}
 		}
-		ms := metadataserver.NewMetaDataServer(context.Background(), backingMetadataServer, iamARN, titusTaskInstanceID, ipv4Address, region, optimistic, container, signer)
+
+		if ipv6Addresses != "" {
```
Comment (sargun) [inline cmd/titus-metadata-service/main.go]: Technically EC2_IPV6s can have multiple IPs. I believe they get newline separated. We only care for the primary IP.

File: cmd/titus-metadata-service/main.go (hunk 2 above)
Comment (rgulewich) [reply cmd/titus-metadata-service/main.go]: Yeah - looks like I didn't push my local change that also did this, sorry. :(

File: metadataserver/iam.go [hunk 4]
```diff
@@ -96,7 +111,18 @@ func newIamProxy(ctx context.Context, router *mux.Router, iamArn, titusTaskInsta
 	*/
 	router.HandleFunc("/security-credentials/{iamProfile}", proxy.specificInstanceProfile)
 
-	if optimistic {
+	if config.APIProtectEnabled {
+		apiProtectInfoLock.Lock()
+		ec2AWSCfg := aws.NewConfig().WithMaxRetries(3)
```
Comment (rgulewich) [inline metadataserver/iam.go]: Why is `WithMaxRetries` set for this AWS config object, but not the STS one?

Comment (sargun) [reply metadataserver/iam.go]: Because the ec2 APIs fail often due to rate limits and such. Whereas we hand-do retries in the STS codepath.

Comment (sargun) [reply metadataserver/iam.go]: It fails open, which is nice as well.

File: cmd/titus-metadata-service/main.go (hunk 3 above)
Comment (rgulewich) [reply cmd/titus-metadata-service/main.go]: Might be helpful to move this comment into the code :)

**Pull request merged**